### PR TITLE
Add preload_paths filter for widgets screen and full site editing

### LINKED ISSUE #28690: [Feature widgets screen] a preload_path filter is missing
## Description
In the block editor use for post we are able to filter the $preload_paths variable.
See https://github.com/WordPress/WordPress/blob/5.6/wp-admin/edit-form-blocks.php#L70-L83

It is usefull:
- to add our own paths to take advantage of the prelaoding process and to retrieve some of our datas to build the UI as soon as possible without making a REST API call.
- to be able to add the same parameter we're injecting with the use of apiFetch.use middleware not to trigger a new REST API request. For example, I think to the language code for our need.

It seems that this kind of filter doesn't exist yet in for the new widget screen nor for the FSE.
See 
https://github.com/WordPress/gutenberg/blob/v9.8.3/lib/widgets-page.php#L57-L66
https://github.com/WordPress/gutenberg/blob/v9.8.3/lib/full-site-editing/edit-site-page.php#L120-L132

## Expected behaviour
Have the same kind of preload_path filter than for the block editor.

### PR DESCRIPTION
**Pull request by manooweb**
## Description
This PR propose to fix https://github.com/WordPress/gutenberg/issues/28690 by adding a filter for widgets screen and full site editing features to be able to preloading datas before the pages is loaded.

**Edit:** I've just updated this PR and the issue with a comment because I saw I had forgotten the navigation editor.

### CODE REVIEW

**Comment by gziolo** [comment]
Thank you for working on it. It feels like it's about time to abstract this editor setup to avoid code duplication but, more importantly, to prevent this type of issue.

**Comment by manooweb** [comment]
Thanks!
Sure! Maybe Gutenberg itself could be use a filter because the paths couldn't be the same in function of the context (editor, widgets-editor...)
However, it also depends on how it will be merged later in the WordPress core.

**Comment by gziolo** [comment]
We could always have someting like:

```php
initialize_block_editor( array(
    'preload_paths' => array(
        // data
    ),
    // more shared settings
) );
```

**Comment by talldan** [comment]
Thanks for working on this. It also closes #24642. We'll need to check that preloading works, last time it was tested it caused some bugs and the implementation didn't go ahead - https://github.com/WordPress/gutenberg/pull/24672.

**Comment by manooweb** [comment]
> Thanks for working on this. Also related is #24642.
> 
> edit: I realise this doesn't add any paths for the navigation editor, just the preloading mechanism, so it doesn't actually resolve the above issue.

Yes! for the navigation editor there was nothing when I worked so I had to implement just the preloading mechanism.
So I can continue to work on this PR:
- to add preload paths suggested in #24642
- improve this part of code to make it common between editors as suggest @gziolo 

Let me know.

**Comment by manooweb** [comment]
> We could always have someting like:
> 
> ```
> initialize_block_editor( array(
>     'preload_paths' => array(
>         // data
>     ),
>     // more shared settings
> ) );
> ```
> 
> Edit: I see there is `gutenberg_get_common_block_editor_settings` already that is a good start.

@gziolo we agree for the block editor, this is the responsability of WordPress core now ?
See https://github.com/WordPress/WordPress/blob/5.6.1/wp-admin/edit-form-blocks.php#L46-L83

The work in progress here could be also merged in WordPress core in a future release, right ?

File: lib/navigation-page.php [hunk 1]
```diff
@@ -40,6 +40,23 @@ function gutenberg_navigation_init( $hook ) {
 	);
 	$settings = gutenberg_experimental_global_styles_settings( $settings );
 
+	// Preload block editor paths.
+	// most of these are copied from edit-forms-blocks.php.
+	$preload_paths = array();
+
+	$preload_paths = apply_filters( 'navigation_editor_preload_paths', $preload_paths );
```
Comment (nosolosw) [inline lib/navigation-page.php:47]: For the navigation screen, I see that there was an attempt to preload certain requests in https://github.com/WordPress/gutenberg/pull/24672#issuecomment-680375627 but it didn't work well with the kind of entities managed by the navigation menu.

It looks to me that this PR could land without this part, which can be iterated later if/when we figure how to preload things for this editor.

**Comment by gziolo** [comment]
We can have a follow-up PR with a new method that would get proposed to the core just after it lands in Gutenberg. That means your assumption is all correct.

**Comment by manooweb** [comment]
As suggested by @gziolo https://github.com/WordPress/gutenberg/pull/28701#issuecomment-773256605 I proprose a modification to initialize each block-based editor the same way.

So I added the `gutenberg_initialize_editor()` function in editor-settings.php to put all is common for each of the editors including a filter for `preload_paths`.

I didn't deal with post block editor because, it seems to me that it isn't similar than the other, especially because of the way it initialized:
-  its name is block_editor in prelaod_path filters name and editor in the initialize function parameter
See https://github.com/WordPress/WordPress/blob/5.6.1/wp-admin/edit-form-blocks.php#L70
and https://github.com/WordPress/WordPress/blob/5.6.1/wp-admin/edit-form-blocks.php#L421
-  there is more than one parameter to initialize the post block editor
See https://github.com/WordPress/WordPress/blob/5.6.1/wp-admin/edit-form-blocks.php#L421 too

I think we can add some other parameters passed to the new gutenberg_initialize_editor() function to be able to specialize more its code to the post block editor need.

For information, I confirm that the preload_paths filter is a real need for us as I described in the initial #28690 issue.

**Comment by manooweb** [comment]
I rebased on trunk and corrected `gutenberg_get_common_block_editor_settings()` documentation to respect coding standards which previously broke the PHP lint check job.

**Comment by manooweb** [comment]
This PR only modify PHP code and doesn't touch anything about React Native.
Is there something I can do with `React Native E2E Tests (Android) ` which break or is there just a problem on CI environment ?

**Comment by gziolo** [comment]
> This PR only modify PHP code and doesn't touch anything about React Native.
> Is there something I can do with `React Native E2E Tests (Android) ` which break or is there just a problem on CI environment ?

It looks like a general issue with the CI job. I see the same failure on `trunk` branch. I asked on WordPress Slack in #mobile channel for clarification (link requires registration at https://make.wordpress.org/chat/):
https://wordpress.slack.com/archives/C02RQC4LY/p1615271529010900

**Comment by hypest** [comment]
👋 all.

> I see the same failure on trunk branch

Hey @gziolo , looking at the logs it looks like the fail here (timing out trying to load some gradle cache) was different than the ones on trunk (`Cannot read property 'stopDriver' of undefined` on a gallery block test).

I restarted the job here since it looks like an intermediate CI infra issue. We can investigate more if it persists with the same error. Sorry for the inconvenience!

File: lib/editor-settings.php [hunk 2]
```diff
@@ -81,3 +81,65 @@ function gutenberg_extend_post_editor_settings( $settings ) {
...
+		'preload_paths'        => array(),
+		'editor_name'          => 'widgets_editor',
+		'editor_script_handle' => 'edit-widgets',
+		'initializer_name'     => 'initialize',
+		'editor_settings'      => array(),
+	);
+
+	$settings = wp_parse_args( $settings, $defaults );
+
+	$preload_paths = apply_filters( "{$settings['editor_name']}_preload_paths", $settings['preload_paths'] );
```
Comment (swissspidy) [inline lib/editor-settings.php:111]: Filter needs documentation.

For readability I'd also do:

```suggestion
	$editor_name = $settings['editor_name'];
	$preload_paths = apply_filters( "{$editor_name}_preload_paths", $settings['preload_paths'] );
```

File: lib/editor-settings.php [hunk 3]
```diff
@@ -81,3 +81,65 @@ function gutenberg_extend_post_editor_settings( $settings ) {
...
+ *      @type array  $editor_settings      editor settings.
+ * }
+ * @return void
+ */
+function gutenberg_initialize_editor( $settings ) {
+
+	$defaults = array(
+		'preload_paths'        => array(),
+		'editor_name'          => 'widgets_editor',
+		'editor_script_handle' => 'edit-widgets',
```
Comment (swissspidy) [inline lib/editor-settings.php:104]: Why is "widgets" the default? Seems quite odd.

Comment (manooweb) [reply lib/editor-settings.php:104]: It is simply my first need . What is preferable then ? Empty strings but it is required in the code behind.

**Comment by manooweb** [comment]
Pushed again by taking in account the latest @swissspidy review comments. Thanks 😉

**Review by draganescu — APPROVED** [review]
This tests well, and the code is a nice refactoring removing useless duplication, while also allowing for customizable preload_paths.  From me, :shipit:

**Comment by gziolo** [comment]
Awesome work @manooweb, this is a great start towards unifying the way block editor is initialized on all pages 👍🏻

**Comment by manooweb** [comment]
Thank you for merging 👍 and your encouragement ☺

**Comment by gziolo** [comment]
I started looking into bringing this new helper into WordPress core. It looks like it might be quite a journey, see https://github.com/WordPress/wordpress-develop/pull/1118 :)

Trac ticket: https://core.trac.wordpress.org/ticket/52920

**Pull request merged**